File: command_generator.rb
=begin

  Module receives JSON output from interpreter, converts series of commands into 
  executable instructions for robot. 

  built in Ruby version 2.1.3p242 ( 2014-09-19 revision 47630 ) [ x86_64-darwin13.0 ]
  requires only standard libraries.

=end

# Libraries for access to the underlying operating system http implementations; included in standard Ruby library
require 'net/http'
require 'uri'
require 'socket'
# library for JSON implementation; included in standard Ruby library
require 'json'

# Class defined for compartmentalization
class CommandGenerator

  # Stardard initialization as part of generating class object; requires JSON file to be interpreted for object creation
  def initialize( data )
    # Assign the opened and parsed JSON object to @object instance variable
    @object = JSON.parse( data )
    # Uncomment for test suite
    # @object = JSON.parse( open( data ).read )
    # Collection of Regexp structures for parsing; any added structures will need to be added in least common to most common order
    # to prevent more common structures from being plucked out of the middle of less common structures
    @structures = [  /((while|for|if|elsif) (x|y) (>|<|>=|<=|==|%|\+|-|\*|\/|\*\*) [0-9]+ (== ([0-9]+|(x|y)))?)/,
                        # catches for, while, and if through a number 0-9 inclusive with optional modulo with arithmatic and subsequent comparison
                    /(((x =)|(y =)) [0-9]+)|(((x =)|(y =)) ... [0-9]+)/,
                        # catches assignments of x/y to a single number 0-9 including x = x + 1 format
                    /((f|b|l|r) (- )?(10|20|30|40|50|60|70|80|90))/,
                        # catches robot commands, units only in tens
                    /(end)|(else)/ ]
                        # catches 'end' and 'else', any and all use cases 
    # Command string to be formatted and passed to robot in the event of non-executable input
    @rescue = "output.push( 'r 10' ); output.push( 'l 10' ); output.push( 'r 10' ); output.push( 'l 10' )"
    # Adds flow control method to the end of initilazation to begin command parsing and generation
    parse_to_astrobot
  end

  # Flow control
  def parse_to_astrobot
    # Pull pertinent data from JSON input ( details in function below )
    pull_commands
    # Break command string into lines ( details in function below )
    parse_string
    # Reassemble and eval command string for output commands ( details in function below )
    eval_string
    # Format into output JSON object ( details in function below )
    json_formatter
  end

  # Pull pertinent data from JSON input; puts all tokens into @raw_command_string
  def pull_commands
    # Array to hold/sort hash values
    hash_holder = []
    # Create instance variable to hold a raw string of commands
    @raw_command_string = ""
    # Go through the JSON object, and for each of the token_values
    @object.each do | key, token |
      # Check to see if the last character added to the @raw_command_string instance variable was part of a multi-digit number
      if hash_holder[ -1 ] && /[0-9]/.match( hash_holder[ -1 ][ 1 ] ) != nil && /[0-9]/.match( token ) != nil
        # Combine last two numbers
        token = "#{ hash_holder[ -1 ][1] }#{ token }"
        # Delete redundant number
        hash_holder.delete_at( -1 )
      end
      # ... add the value to the @raw_command_string instance variable
      hash_holder << [ key, token ] if  token != nil || token != ""
    end
    # Sort array by key
    hash_holder.sort { | a, b| a[ 0 ] <=> b[ 0 ] }.each do | pair |
      # Delete index from final array
      pair.delete_at( 0 )
    end
    # Convert array to formatted string
    @raw_command_string = hash_holder.flatten.join( " " )
  end

  # Break command string into lines
  def parse_string
    # Local array ( multidimensional ) to hold code structures
    parse_string_holder = []
    # Loop through each of the regular expressions in the @structures instance variable array
    @structures.each do | structure |
      # Check to see if the structure can be found in the @raw_command_string created in the previous method call
      if structure.match( @raw_command_string )
        # If any were found, loop through the found objects
        @raw_command_string.scan( structure ).each do | caught |
          # Pull out the longest array element corresponding to the whole found string
          # this is necessary due to the way Ruby handles results from a regex search; creating a MatchData
          # object which can be read and treated as an array
          found = caught.to_a.compact.max_by( &:length )
          # Find the starting point in the @raw_command_string instance variabl array using the found string as the
          # basis for a new regex.
          start = /#{ Regexp.quote( found ) }/.match( @raw_command_string ).begin( 0 )
          # Holder for substitution string
          substitute = ""
          # Create substitution string using least common character I could think of, if it proves necessary, this can
          # be swapped out for any other single character in the event "~" is needed down the road
          found.length.times { substitute << "~" }
          # Replace the structure in the original string ( ! = dustructive, meaning it changes the orignal object ) 
          # with the substitute string to prevent duplication
          @raw_command_string.sub!( structure,  substitute )
          # Add subarray to holder array consisting of the starting point ( for sorting ) and the found string
          parse_string_holder << [ start, found ]
        end
      end
    end
    # Check for non-valid structures, throw error
    if @raw_command_string.gsub( /~/, "" ).gsub( / /, "" ).length > 0
      # If there are any uncaught ( invalid ) block structures, set to false for error handling
      @parsed_command_string = false
    else
      # Assign to new instance variable a sorted multidimensional array consisting of all found command structures in 
      # order of execution
      @parsed_command_string = parse_string_holder.sort
    end
  end

  # Reassemble and eval command string for output commands
  def eval_string
    # Local holder for executable code strings
    eval_string_holder = []
    # Local holder for final JSON output code
    output = []
    # Check for invalid command sequence from above
    if @parsed_command_string
      # Go through the @parsed_command_string instance variable created above
      @parsed_command_string.each do | subcommand |
        # Checks to see if the commands are intended for the robot
        if subcommand[ 1 ].match( /((f|b|l|r) (10|20|30|40|50|60|70|80|90))/ )
          # Adds internal code to enable robot instructions to be inserted and appropriately formatted into the local output
          # holder array
          subcommand[ 1 ] = "output.push( '#{ subcommand[ 1 ] }' )"
        end
        # Adds code strings to be executed to local holder array
        eval_string_holder << subcommand[ 1 ]
      end
      # Combines the disparate bits of code with ";" line separation character
      executable = eval_string_holder.join( "; " )
      # Rescue loop
     begin
        # Execute code string
        eval( executable )
      # If there's an error...
     rescue Exception => exc
        # Execute the @rescue string instead ( defined in initilazation )
       eval( @rescue )
     end
   else
    # Beep
    print "\a"
    # Rescue sequence
    eval( @rescue )
  end
    # Assign the contents of the output holder array ( formatted robot commands )to new @command_array 
    # instance variable for access below
    @command_array = output
  end

  # Format into output JSON object
  def json_formatter
    # Create hash formatted to robot's API for final output with empty "command_values" array
    # json_output = { "robot_commands" => @command_array.length, "command_values" => [] }
    json_output = { }
    # Loop through the @command_array instance variable created above
    command_body = @command_array.each do | command |
      # Add a hash to the "command_values" array in the json_output hash for each robot instruction
      # json_output[ "command_values" ] << { "line" => "#{ @command_array.index( command ) }", "value" => "#{ command }" } 
      json_output[ "cmd#{ @command_array.index( command ) + 1 }" ] = "#{ command }" 
      # Set the array element to empty string so that the proper index is set above when it comes back around
      # in the event of duplicate instructions
      @command_array[ @command_array.index( command ) ] = ""
    end 
    # Convert json_output hash to JSON object, and assign to @output new instance variable
    @output = json_output.to_json
  end

end

class AstroBotServer


  IP_address_in = "192.168.43.153"
  port_in = 9999
  IP_address_out = "192.168.43.36"
  port_out = 8081

  server = TCPServer.new( IP_address_in, port_in )
  # server = TCPServer.new( "10.2.108.1", 9999 )
  STDERR.puts "Command Generator active."
  STDERR.puts "Address = http://#{IP_address_in}:#{port_in}"

  # Servers run forever
  loop {
    Thread.start( server.accept ) do | client |
      # Receive section
      # In this case, method = "POST" and path = "/"
      method, path = client.gets.split
      headers = {}
      # Collect HTTP headers
      while line = client.gets.split( ' ', 2 )
        # Blank line means no more headers
        break if line[ 0 ] == ""
        # Hash headers by type
        headers[ line[ 0 ].chop ] = line[ 1 ].strip
      end
      # Read the POST data as specified in the header
      data = client.read( headers[ "Content-Length" ].to_i )
      STDERR.puts "Connection received."
      STDERR.puts "Input @#{Time.now}:"
      STDERR.puts data
      response = "HTTP/1.1 200 OK\r\nContent-type: application/json\r\n\r\n"
      client.puts response
      client.close
      # Send section

      url1 = "http://#{IP_address_out}:#{port_out}/update"
      uri1 = URI.parse( url1 )
      http = Net::HTTP.new( uri1.host, uri1.port )
      payload = CommandGenerator.new( data ).instance_variable_get( :@output )
      request = Net::HTTP::Post.new( uri1.request_uri, initheader = { 'Content-Type' => 'application/json' } )
      request.body = payload
      resp = http.request( request )
      STDERR.puts "Output @#{Time.now}:"
      STDERR.puts payload
    end
  }
end

AstroBotServer.new()

File: command_generator.py
import json
import re
from operator import itemgetter
import http.server
import http.client
import itertools as IT


STRUCTURES_STR = [
    # catches for, while, and if through a number 0-9 inclusive with optional
    # modulo with arithmatic and subsequent comparison
    r'((while|for|if|elsif) (x|y) (>|<|>=|<=|==|%|\+|-|\*|\/|\*\*) [0-9]+ (== ([0-9]+|(x|y)))?)',

    # catches assignments of x/y to a single number 0-9 including x = x + 1
    # format
    r'(((x =)|(y =)) [0-9]+)|(((x =)|(y =)) ... [0-9]+)',

    # catches robot commands, units only in tens
    r'((f|b|l|r) (- )?(10|20|30|40|50|60|70|80|90))',

    # catches 'end' and 'else', any and all use cases
    r'(end)|(else)',
]

# Now we take all the regexs defined in STRUCTURES_STR and combine them into
# one compiled regex that can match for all the structures
STRUCTURES = re.compile(r'({})'.format('|'.join(STRUCTURES_STR)))


class InvalidCommand(Exception):
    """
    Exception for invalid commands in the command generator
    """
    def __init__(self, command, line=None):
        self.command = command
        self.line = line

    def __str__(self):
        if self.line:
            return "Invalid command on line {}: {}".format(
                self.line, self.command
            )
        return "Invalid command: {}".format(self.command)


class CommandGenerator(object):
    # Command string to be formatted and passed to robot in the event of
    # non-executable input
    FAIL = "r(10) l(10) r(10) l(10)".split(' ')

    def __init__(self, data):
        try:
            data_json = json.loads(data)
            commands = self._parse_commands(data_json)
            structures = self._parse_structure(commands)
            output = self._evaluate(structures)
        except Exception as e:
            print(e)
            output = self._evaluate(self.FAIL)
        finally:
            self.output = self._parse_output(output)

    def _parse_commands(self, translator_output):
        # sort translator output by the key value
        sorted_commands = sorted(
            translator_output.items(), 
            key=lambda item: int(item[0])
        )
        # extract the commands and filter out false-ish values
        commands_raw = filter(None, map(itemgetter(1), sorted_commands))
        # group sequences of numbers in the commands
        commands = []
        for is_numbers, command_group in IT.groupby(commands_raw, str.isdigit):
            if is_numbers:
                commands.append("".join(command_group))
            else:
                commands.extend(command_group)
        return " ".join(commands)

    def _parse_structure(self, commands):
        # structures start as a mapping from "start index of structure" =>
        # "structure code"
        structures = {}
        # leftovers will get re-written with '#' to mark sections we have
        # already found
        leftover = commands
        # use the STRUCTURES regex to find all structures in the inputted
        # command
        for match in STRUCTURES.finditer(commands):
            start, end = match.span(0)
            structures[start] = match.group(0).strip()
            leftover = leftover[:start] + '#' * (end-start) + leftover[end:]
        # if leftovers is only # and spaces then it only contains valid
        # structures
        leftover = leftover.strip("# ")
        if leftover:
            raise InvalidCommand(leftover)
        # sort the structures on their starting indexes from the original
        # commands string
        structures_sorted = sorted(structures.items(), key=itemgetter(0))
        return map(itemgetter(1), structures_sorted)

    def _translate_structures(self, commands):
        """
        The following is terrible code that only god can forgive us for.  It
        translate the pseudo-code recived from the translator into python.
        """
        structure = []
        depth = 0
        for i, command in enumerate(commands):
            # in python, 'elsif' is 'elif'
            command = command.replace("elsif", "elif")
            if command[:2].isalpha():
                # append collon at end of control structures
                command += ":"
            elif any(command.startswith(f + ' ') for f in 'fblr'):
                # turn calls to f,b,l and r into actual function calls
                command = "{}({})".format(command[0], command[1:])
            if command.startswith('end'):
                # 'end' structure decreases the python depth
                depth -= 1
                # we skip the rest of the loop to skip adding this to the
                # structure list
                continue
            elif any(command.startswith(f) for f in ('else', 'elif')):
                # 'else' and 'elif' structure decreases the python depth
                depth -= 1
            # add the current command at the correct depth
            structure.append("\t"*depth + command)
            if any(command.startswith(f) for f in
                    ('if', 'while', 'for', 'else', 'elif')):
                # anything after a 'if', 'while', 'for', 'else' or 'elif'
                # should be at an increased depth
                depth += 1
        return structure

    def _evaluate(self, structures):
        # translate structures into python
        structures = self._translate_structures(structures)
        # create an output array and pin calls to f/l/r/b into appends in this
        # output array
        output = []
        eval_globals = {
            'f': lambda x: output.append('f ' + str(x)),
            'l': lambda x: output.append('l ' + str(x)),
            'r': lambda x: output.append('r ' + str(x)),
            'b': lambda x: output.append('b ' + str(x)),
        }
        # execute the structures in a controlled environment that only contains
        # our patched f/l/r/b functions
        exec("\n".join(structures), eval_globals, {})
        return output

    def _parse_output(self, output):
        return {'cmd'+str(i): cmd for i, cmd in enumerate(output, 1)}

    def __str__(self):
        return json.dumps(self.output)


class HTTPHandler(http.server.BaseHTTPRequestHandler):
    output_server = None

    def do_POST(s):
        """
        Read the data from the post request (in UTF-8 encoding) and pushes the
        result to the server given by the 'output_server' class variable
        """
        print("Processing request")
        data_length = int(s.headers['Content-Length'])
        data = s.rfile.read(data_length).decode("utf-8")
        commands = CommandGenerator(data)

        s.send_response(200)
        s.send_header("Content-type", "application/json")
        s.end_headers()

        conn = http.client.HTTPConnection(*s.output_server)
        conn.request("PUT", "/update", str(commands))
        conn.close()


if __name__ == "__main__":
    input_server = ("127.0.0.1", 9999)
    output_server = ("127.0.0.1", 8081)

    HTTPHandler.output_server = output_server
    httpd = http.server.HTTPServer(input_server, HTTPHandler)
    print("Command Generator active.")
    print("Address = http://{}:{}".format(*input_server))
    httpd.serve_forever()
